Add Google VertexAI support.

steveheyman opened the pull request:
Support VertexAI framework.
Support detection and error handling.
This pull request closes issue #33

liushuangls commented on client.go:
```diff
@@ -74,13 +74,54 @@ func (c *Client) handlerRequestError(resp *http.Response) error {
+				return &reqErr
+			}
+
 		return fmt.Errorf("error, status code: %d, message: %w", resp.StatusCode, errRes.Error)
```
Notice the indentation here

steveheyman replied:
Fixed

liushuangls commented on vertexerror.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+/*
+func (e *VertexAPIError) IsInvalidRequestErr() bool {
```
These are not needed and can be deleted directly

steveheyman replied:
Done

liushuangls commented:
Thanks to your PR, I thought we should abstract out an interface that different api platforms just need to implement instead of changing the code when adding a new platform

WillMatthews commented on README.md:
```diff
@@ -157,7 +157,7 @@ func main() {
 		return
 	}
-	fmt.Println(resp.Content[0].Text)
+	fmt.Println(*resp.Content[0].Text)
```
Prefer a `.GetText()` to a direct dereference.

steveheyman replied:
Fixed

WillMatthews commented on README.md:
```diff
@@ -305,6 +305,77 @@ func main() {
+		}
+		return
+	}
+	fmt.Println(*resp.Content[0].Text)
```
Again, prefer a `.GetText()` to direct dereference.

steveheyman replied:
Fixed

WillMatthews commented on vertexerror.go:
```diff
@@ -0,0 +1,60 @@
+	} `json:"metadata"`
+}
+
+/*
```
Please delete these if they are not used.

steveheyman replied:
Done

WillMatthews commented on common.go:
```diff
@@ -21,3 +21,22 @@ const (
 	RoleAssistant ChatRole = "assistant"
 )
+
+func translateVertexModel(model Model) string {
```
I'd prefer to see this as a method on `Model`, i.e.
```suggestion
func (m Model) asVertex() string {
```

steveheyman replied:
Fixed

WillMatthews commented on config.go:
```diff
@@ -87,3 +92,19 @@ func WithBetaVersion(betaVersion ...BetaVersion) ClientOption {
+
+func WithVertexAI(projectID string, location string) ClientOption {
+	return func(c *ClientConfig) {
+		c.BaseURL = fmt.Sprintf("https://%s-aiplatform.googleapis.com/v1/projects/%s/locations/%s/publishers/anthropic/models", location, projectID, location)
```
I feel like we might be able to do a better job here using `net/url` to build this url (although I'm sure the `Sprintf` is probably fine for now).

steveheyman replied:
I considered using url.JoinPath, but it returns an err.  Since the func has no way to return an err, it would either need to be silently ignored, or panic, neither of which feels like a great option.  However, happy to fix to use this approach.  I did see that client.go line 121 uses Sprintf so it is not unprecedented.

WillMatthews commented on config.go:
```diff
@@ -87,3 +92,19 @@ func WithBetaVersion(betaVersion ...BetaVersion) ClientOption {
+	return func(c *ClientConfig) {
+		c.apiKeyFunc = apiKeyFunc
+	}
+}
+func isVertexAI(apiVersion APIVersion) bool {
```
Whitespace
```suggestion
}

func isVertexAI(apiVersion APIVersion) bool {
```

steveheyman replied:
Fixed

WillMatthews commented on config.go:
```diff
@@ -87,3 +92,19 @@ func WithBetaVersion(betaVersion ...BetaVersion) ClientOption {
+	return func(c *ClientConfig) {
+		c.apiKeyFunc = apiKeyFunc
+	}
+}
+func isVertexAI(apiVersion APIVersion) bool {
```
I think this could be a method on the `APIVersion` type. (maybe call it `isVertex()`)

steveheyman replied:
Done

WillMatthews commented on client.go:
```diff
@@ -125,8 +177,18 @@ func (c *Client) newRequest(
-	req.Header.Set("Anthropic-Version", string(c.config.APIVersion))
+
+	apiKey := c.config.apiKey
+	if c.config.apiKeyFunc != nil {
+		apiKey = c.config.apiKeyFunc()
+	}
```
What's the purpose of the `apiKeyFunc` out of curiosity?

WillMatthews commented on client.go:
```diff
@@ -105,6 +146,17 @@ func (c *Client) newRequest(
+			model = vertexAISupport.GetModel()
+			vertexAISupport.SetAnthropicVersion(c.config.APIVersion)
+		} else {
+			return nil, fmt.Errorf("this call not supported by the Vertex AI API")
```
```suggestion
			return nil, fmt.Errorf("this call is not supported by the Vertex AI API")
```

steveheyman replied:
Fixed

WillMatthews reviewed (commented):
Some initial comments. Would it be possible to have an integration test?

steveheyman commented on client.go:
```diff
@@ -74,13 +74,54 @@ func (c *Client) handlerRequestError(resp *http.Response) error {
+			}
+			return fmt.Errorf("error, status code: %d, message: %w", resp.StatusCode, errRes.Error)
+		} else {
+			var errRes ErrorResponse
```
This was a merge error on my part.  I removed the duplicate code.

steveheyman commented:
I addressed all of your comments with my latest check-ins.  I added an integration test, which requires setting the VERTEX_API_KEY as well as the VERTEX_LOCATION and VERTEX_PROJECT, which would need to be created through GCloud.  One small note, the integration test requires a dependency on "golang.org/x/oauth2/google".
)

liushuangls commented on README.md:
```diff
@@ -157,7 +157,7 @@ func main() {
 		return
 	}
-	fmt.Println(resp.Content[0].Text)
+	fmt.Println(*resp.Content[0].GetText())
```
No need for * here

liushuangls commented:
> Thanks to your PR, I thought we should abstract out an interface that different api platforms just need to implement instead of changing the code when adding a new platform

I feel that the current solution is more invasive to the code, if we later add a new api provider (such as aws), it may also add a lot of if else judgment, the better solution is to abstract a common interface, what do you think? @WillMatthews

steveheyman commented:
Ok, I just refactored everything to use an adapter.  The adapter can contain provider-specific code, while keeping the existing Client implementation handling most of the common logic.  The base code is much cleaner now, and should be fairly extensible to AWS or other providers.

liushuangls commented:
Thank you very much!

liushuangls commented:
Please modify the golines format check error

steveheyman commented:
Done.  I also added some unit tests for Vertex specific code to fix the code coverage.

WillMatthews commented:
> > Thanks to your PR, I thought we should abstract out an interface that different api platforms just need to implement instead of changing the code when adding a new platform
> 
> I feel that the current solution is more invasive to the code, if we later add a new api provider (such as aws), it may also add a lot of if else judgment, the better solution is to abstract a common interface, what do you think? @WillMatthews

Whoops - sorry for the slow reply. I agree that an interface is the right approach here. Thank you very much for your work on this PR @steveheyman! It's exciting to now have VertexAI support built in :smiley:
